File: 0290-word-pattern/0290-word-pattern.js
/**
 * @param {string} pattern
 * @param {string} s
 * @return {boolean}
 */
var wordPattern = function(pattern, s) {
    const sMap = new Map();
    const pMap = new Map();

    sWords = s.split(" ");
    pChars = pattern.split("");
    
    if (pChars.length != sWords.length) {
        return false;
    }
    for (let i = 0; i < sWords.length; i++) {
        const pChar = pChars[i];
        const sWord = sWords[i];

        if ((sMap.has(sWord) && sMap.get(sWord) != pChar)
            || (pMap.has(pChar) && pMap.get(pChar) != sWord)) {
            return false;
        }

        sMap.set(sWord, pChar);
        pMap.set(pChar, sWord);
    }
    return true;
};

File: 0290-word-pattern/0290-word-pattern.java
class Solution {
    public boolean wordPattern(String pattern, String s) {
        Map<Character, String> pMap = new HashMap<>();
        Map<String, Character> sMap = new HashMap<>();

        String[] words = s.split(" ");
        if (pattern.length() != words.length) return false;

        for (int i = 0; i < pattern.length(); i++) {
            char c = pattern.charAt(i);
            String word = words[i];
            if ((pMap.containsKey(c) && !pMap.get(c).equals(word)) ||
                (sMap.containsKey(word) && !sMap.get(word).equals(c))) {
                return false;
            }
            if (!pMap.containsKey(c)) {
                pMap.put(c, word);
            }
            if (!sMap.containsKey(word)) {
                sMap.put(word, c);
            }
        }
        return true;
    }
}